File: src/arduinobot_controller/scripts/angles_converter.py
#!/usr/bin/env python3
import rospy
from arduinobot_controller.srv import AnglesConverter, AnglesConverterResponse
import math

"""
Function call back for radians_to_degrees service. This function receives the position for the
arm in radians and converts them to degrees.
"""
def convert_radians_to_degrees(req):
    res = AnglesConverterResponse()
    res.base = int(((req.base+(math.pi/2))*180)/math.pi)
    res.shoulder = 180-int(((req.shoulder+(math.pi/2))*180)/math.pi)
    res.elbow = int(((req.elbow+(math.pi/2))*180)/math.pi)
    res.gripper = int(((-req.gripper)*180)/(math.pi/2))

    return res

"""
Function call back for degrees_to_radians service. This function receives the position for the
arm in degrees and converts them to radians.
"""
def convert_degrees_to_radians(req):
    res = AnglesConverterResponse()
    res.base = ((math.pi*req.base) - ((math.pi/2)*180))/180
    res.shoulder = (((180-req.shoulder)*math.pi)-((math.pi/2)*180))/180
    res.elbow = ((math.pi*req.elbow) - ((math.pi/2)*180))/180
    res.gripper = -((math.pi/2)*req.gripper)/180

    return res

if __name__ == '__main__':
    # Initialize a ROS node names angles_converter
    rospy.init_node("angles_converter")

    # Initialize a service to convert radians to degrees and vice versa
    radians_to_degress = rospy.Service("radians_to_degress", AnglesConverter, convert_radians_to_degrees)
    degrees_to_radians = rospy.Service("degress_to_radians", AnglesConverter, convert_degrees_to_radians)
    
    # keeps the node up and running
    rospy.spin()

File: src/arduinobot_controller/src/angles_converter.cpp
#include <ros/ros.h>
#include "arduinobot_controller/AnglesConverter.h"
#include <math.h>

bool convertRadiansToDegrees(arduinobot_controller::AnglesConverter::Request &req,
                             arduinobot_controller::AnglesConverter::Response &res)
{
    res.base = static_cast<int>(((req.base + (M_PI / 2)) * 180) / M_PI);
    res.shoulder = 180 - static_cast<int>(((req.shoulder + (M_PI / 2)) * 180) / M_PI);
    res.elbow = static_cast<int>(((req.elbow + (M_PI / 2)) * 180) / M_PI);
    res.gripper = static_cast<int>(((-req.gripper) * 180) / (M_PI / 2));

    return true;
}

bool convertDegreesToRadians(arduinobot_controller::AnglesConverter::Request &req,
                             arduinobot_controller::AnglesConverter::Response &res)
{
    res.base = ((M_PI * req.base) - ((M_PI / 2) * 180)) / 180;
    res.shoulder = (((180 - req.shoulder) * M_PI) - ((M_PI / 2) * 180)) / 180;
    res.elbow = ((M_PI * req.elbow) - ((M_PI / 2) * 180)) / 180;
    res.gripper = -((M_PI / 2) * req.gripper) / 180;

    return true;
}

int main(int argc, char **argv)
{

    ros::init(argc, argv, "angles_converter");
    ros::NodeHandle n;
    ros::ServiceServer radiansToDegrees = n.advertiseService("radians_to_degrees", convertRadiansToDegrees);
    ros::ServiceServer degreesToRadians = n.advertiseService("degrees_to_radians", convertDegreesToRadians);

    ros::spin();
    return 0;
}